larq/optimizers_v1.py

```diff
@@ -0,0 +1,180 @@
+
+        if isinstance(model, tf.keras.Model):
+            self.multipliers = {}
+            for layer in model.layers:
+                if hasattr(layer, "quantized_latent_weights"):
+                    for weight in layer.quantized_latent_weights:
+                        self.multipliers[weight.name] = self.get_lr_multiplier(weight)
+        elif isinstance(model, dict):
+            self.multipliers = model
+        else:
+            raise ValueError(f"Expected tf.keras.Model or dict, received {type(model)}")
+
+    def get_lr_multiplier(self, weight):
+        shape = weight.get_shape().as_list()
+        n_input = shape[-2]
+        n_output = shape[-1]
+        if len(shape) == 4:
+            kernelsize = np.prod(shape[:-2])
+            coeff = 1.0 / np.sqrt(1.5 / ((kernelsize * (n_input + n_output))))
+        elif len(shape) == 2:
+            coeff = 1.0 / np.sqrt(1.5 / ((1.0 * (n_input + n_output))))
+        else:
+            raise NotImplementedError(
+                "Xavier Learning rate scaling not implimented for this kernelsize"
+            )
+        return coeff
+
+    def get_updates(self, loss, params):
+        mult_lr_params = [p for p in params if p.name in self.multipliers]
+        base_lr_params = [p for p in params if p.name not in self.multipliers]
+
+        updates = []
+        base_lr = self.optimizer.lr
+        for param in mult_lr_params:
+            self.optimizer.lr = base_lr * self.multipliers[param.name]
+            updates.extend(self.optimizer.get_updates(loss, [param]))
+
+        self.optimizer.lr = base_lr
+        updates.extend(self.optimizer.get_updates(loss, base_lr_params))
+
+        return updates
+
+    def __getattr__(self, name):
+        return getattr(self.optimizer, name)
+
+    def get_config(self):
+        return {
+            "optimizer": {
+                "class_name": self.optimizer.__class__.__name__,
+                "config": self.optimizer.get_config(),
+            },
+            "multipliers": self.multipliers,
+        }
+
+    @classmethod
+    def from_config(cls, config, custom_objects=None):
+        optimizer = tf.keras.optimizers.deserialize(
+            config.pop("optimizer"), custom_objects=custom_objects
+        )
+        return cls(optimizer, config["multipliers"])
+
+
+@utils.register_keras_custom_object
+class Bop(tf.keras.optimizers.Optimizer):
+    def __init__(self, fp_optimizer, threshold=1e-5, gamma=1e-2, name="Bop", **kwargs):
+        super().__init__(**kwargs)
+
+        if not isinstance(fp_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(
+                f"Expected tf.keras.optimizers.Optimizer, received {type(fp_optimizer)}."
+            )
+
+        with K.name_scope(self.__class__.__name__):
+
+            self.fp_optimizer = fp_optimizer
+            self.threshold = threshold
+            self.gamma = gamma
+            self.iterations = K.variable(0, dtype="int64", name="iterations")
```

Comment:
I don't think `self.iterations` is used anywhere. So we can remove it:
```suggestion
```

Reply:
I am using iterations in test cases.

Reply:
> I am using iterations in test cases.

If it is only used in the test case, I think this shouldn't be here unless all Keras optimizers are expected to have an `iterations` attribute.